### PR TITLE
Use `gradle-git-versions` and support gradle 9

### PR DESCRIPTION
**Pull request by FinlayRJW**
## Before this PR

`GitVersionUtils` shelled out to git via `project.exec`, which is removed in Gradle 9. This prevented the plugin from running on Gradle 9 daemons.

## After this PR

The custom git utilities are replaced by `com.palantir.gradle.gitversion:gradle-git-version` 5.0.0, whose `GitInvoker` runs git through the Gradle Provider API and is Gradle-9-compatible. The plugin is now tested against Gradle 8.14.5 and 9.3.1.

`GitVersionUtils` → `GitOperations` — abstract class wired into `RevapiExtension` via `@Nested`, with the same behaviour: walk back from `HEAD` collecting the previous three tags, filter the initial `0.0.0` tag, strip `v` prefixes.

## What else is in here

This PR also folds in two adjacent cleanups that were needed to get Gradle 9 green:

- **Bytecode target raised from Java 8 → Java 17.** Gradle 9 requires JDK 17 to run the daemon, so the previous Java 8 target was no longer useful. `libraryTarget = 17`, `javaCompiler = 25`, `daemonTarget = 21`.
- **Final Groovy → JUnit test migration** (commit 6a36fc6) — removes the last four `*Spec.groovy` files and the `groovy` plugin from `build.gradle`. Continuation of #683 and #707.

### CODE REVIEW

File: src/main/java/com/palantir/gradle/revapi/GitOperations.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                .orElse(Optional.empty());
+    }
+
+    private Provider<Boolean> commitExists(String ref) {
+        return getGitInvoker()
+                .invokeWithResult("cat-file", "-t", ref)
+                .map(result -> result.uncheckedStandardOut().equals("commit"));
+    }
+
+    private Provider<Optional<String>> describeTagAt(String ref) {
```
Comment (felixdesouza) [inline src/main/java/com/palantir/gradle/revapi/GitOperations.java]: it's a little weird to have an `Optional` type inside something that can also be `Optional`. I would stick to just using the `Provider`.

Concretely, returning `null`, instead of `Optional.empty`.

File: src/main/java/com/palantir/gradle/revapi/ImmutablesFilter.java [hunk 2]
```diff
@@ -72,23 +72,14 @@ public Difference transform(
...
-            case METHOD_REMOVED:
-                return inImmutablesClass(oldElement) && abstractNonPublic(oldElement);
-
-            case METHOD_NOW_ABSTRACT:
-                return inImmutablesClass(oldElement) && inImmutablesClass(newElement);
-
-            default:
-                return false;
-        }
+        return switch (Code.fromCode(difference.code)) {
```
Comment (felixdesouza) [inline src/main/java/com/palantir/gradle/revapi/ImmutablesFilter.java:75]: fantastic

File: build.gradle [hunk 3]
```diff
@@ -78,8 +81,16 @@ allprojects {
     }
 
     group 'com.palantir.gradle.revapi'
-    version System.env.CIRCLE_TAG ?: gitVersion()
-    sourceCompatibility = 1.8
+    version rootProject.version
```
Comment (felixdesouza) [inline build.gradle]: I don't think we need to bother saving `gitVersion()` to a field anymore since it should be cached now?

File: build.gradle [hunk 4]
```diff
@@ -108,21 +114,6 @@ gradlePlugin {
...
-tasks.named('processResources').configure {
-    duplicatesStrategy 'include'
-}
-
-tasks.publish.dependsOn tasks.publishPlugins
-publishPlugins.onlyIf {
-    versionDetails().isCleanTag
-}
-project.ext.'gradle.publish.key' = System.env["GRADLE_KEY"]
-project.ext.'gradle.publish.secret' = System.env["GRADLE_SECRET"]
```
Comment (felixdesouza) [inline build.gradle:124]: do we not need these anymore?

Comment (FinlayRJW) [reply build.gradle:124]: Ah I assumed we didn't as the plugin I was copying from didn't apply - but I think we still do looking at the other plugins that use `com.palantir.external-publish-gradle-plugin`

File: src/main/java/com/palantir/gradle/revapi/GitOperations.java (hunk 1 above)
Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java]: fixed in [72ec406](https://github.com/palantir/gradle-revapi/pull/714/commits/72ec4069f2e1e5ad11106c699b5fd8d5561c036b)

File: build.gradle (hunk 4 above)
Comment (felixdesouza) [reply build.gradle:124]: I don't actually know, I haven't looked at publishing for external gradle plugins in forever, I would also imagine that external-publish-gradle-plugin would sort this out, so it's probably worth double checking whether we can acutally get rid of it

File: build.gradle (hunk 3 above)
Comment (FinlayRJW) [reply build.gradle]: fixed in [0f026ef](https://github.com/palantir/gradle-revapi/pull/714/commits/0f026ef1b2294f35a498da9db4a9abacc663d024)

File: build.gradle (hunk 4 above)
Comment (FinlayRJW) [reply build.gradle:124]: fixed in [0f026ef](https://github.com/palantir/gradle-revapi/pull/714/commits/0f026ef1b2294f35a498da9db4a9abacc663d024)

Comment (FinlayRJW) [reply build.gradle:124]: cool I'll leave as was for now then check and fix in a FLUP if can be removed

Comment (FinlayRJW) [reply build.gradle:124]: Ah will remove now this is dead code - https://github.com/palantir/gradle-external-publish-plugin/pull/684

File: src/main/java/com/palantir/gradle/revapi/GitOperations.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    private Provider<String> previousGitTagFromRef(String ref) {
+        String beforeLastRef = ref + "^";
+        return commitExists(beforeLastRef).flatMap(exists -> {
+            if (!exists) {
+                return getProviderFactory().provider(() -> null);
+            }
+            Provider<String> tag = describeTagAt(beforeLastRef);
+            return tag.flatMap(rawTag -> "0.0.0".equals(rawTag) ? keepTagIfHasParent("0.0.0") : tag);
+        });
+    }
```
Comment (felixdesouza) [inline src/main/java/com/palantir/gradle/revapi/GitOperations.java]: I'm not super keen on the recursion and a lot of the flatmaps, I think we can simplify some of that

File: src/main/java/com/palantir/gradle/revapi/GitOperations.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    private Provider<String> previousGitTagFromRef(String ref) {
+        String beforeLastRef = ref + "^";
+        return commitExists(beforeLastRef).flatMap(exists -> {
+            if (!exists) {
+                return getProviderFactory().provider(() -> null);
+            }
+            Provider<String> tag = describeTagAt(beforeLastRef);
+            return tag.flatMap(rawTag -> "0.0.0".equals(rawTag) ? keepTagIfHasParent("0.0.0") : tag);
+        });
```
Comment (felixdesouza) [inline src/main/java/com/palantir/gradle/revapi/GitOperations.java]: I would change this to:

```java
        Provider<String> tag = describeTagAt(beforeLastRef);
        Provider<String> correctedPreviousGitTag =
                tag.flatMap(rawTag -> "0.0.0".equals(rawTag) ? keepTagIfHasParent("0.0.0") : tag);
        return commitExists(beforeLastRef)
                .zip(correctedPreviousGitTag, (exists, correctedTag) -> !exists ? null : correctedTag);
```

`commitExists` and `describeTagAt` do not depend on each other

File: src/main/java/com/palantir/gradle/revapi/GitOperations.java (hunk 5 above)
Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java]: yeah I wasn't sure the best way todo the three levels of back checking without lots of gets otherwise - not sure if ive missed something though

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java]: okay gone with:

```java
    public final Provider<List<String>> previousGitTags() {
        return previousGitTagFromRef("HEAD")
                .map(seed -> Stream.iterate(
                                seed,
                                Objects::nonNull,
                                ref -> previousGitTagFromRef(ref).getOrNull())
                        .limit(TAGS_TO_RETURN)
                        .map(GitOperations::stripVFromTag)
                        .toList())
                .orElse(List.of());
    }

    private Provider<String> previousGitTagFromRef(String ref) {
        String beforeLastRef = ref + "^";
        return commitExistsAt(beforeLastRef)
                .zip(describeReleaseTagAt(beforeLastRef), (exists, releaseTag) -> exists ? releaseTag : null);
    }
```

avoid recursion and means we don't call unwrap and re-wrap the git provider - lmk if this looks good

File: src/main/java/com/palantir/gradle/revapi/GitOperations.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    private Provider<List<String>> previousGitTagsFromRef(String ref, int remaining) {
+        if (remaining == 0) {
+            return getProviderFactory().provider(List::of);
+        }
+        return previousGitTagFromRef(ref)
+                .flatMap(tag -> previousGitTagsFromRef(tag, remaining - 1)
+                        .map(previousTags -> Stream.concat(Stream.of(stripVFromTag(tag)), previousTags.stream())
+                                .toList()))
+                .orElse(List.of());
+    }
```
Comment (felixdesouza) [inline src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: Thinking some more, I would probably do something along these lines:

```java
    public final Provider<List<String>> previousGitTags() {
        Provider<Builder<String>> seed = getProviderFactory().provider(ImmutableList::builder);
        return StreamEx.iterate(previousGitTagFromRef("HEAD"), this::previousGitTagFromRefProvider)
                .limit(TAGS_TO_RETURN)
                .foldLeft(seed, GitOperations::combine)
                .map(Builder::build);
    }

    private static Provider<ImmutableList.Builder<String>> combine(
            Provider<ImmutableList.Builder<String>> currentTags, Provider<String> next) {
        return currentTags.zip(
                next.orElse(""), (tags, nextTag) -> nextTag.isBlank() ? tags : tags.add(stripVFromTag(nextTag)));
    }

    private Provider<String> previousGitTagFromRefProvider(Provider<String> ref) {
        return ref.flatMap(this::previousGitTagFromRef);
    }
```

The control flow is still a little complicated due to the nested nature of it all, but I don't think I need to reason about recursion as much, it's more:
* keep finding as many tags as possible
* limit to how many we want
* combine them into a list

Comment (felixdesouza) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: actually, this is probably silly. We should just ask git for this!

```
git tag --merged HEAD --sort=-version:refname
```

gets you the tags that are ancestors of `HEAD` which is what you want here, you do get the whole thing however.

It might be beneficial to try this instead:

(reverse sorted by tag creation)
```
git for-each-ref --count=10 --sort=-creatordate --format='%(refname:short)' refs/tags --merged HEAD
```

or (reverse sorted by version)
```
git for-each-ref --count=10 --sort=-version:refname --format='%(refname:short)' refs/tags --merged HEAD
```

This is far simpler than the gymnastics we're having to do here

Comment (felixdesouza) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: we just need to ensure we have the same behaviour for whatever the `0.0.0` behaviour is.

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: Yeah I was going this route initially but was worried that the ordering might not be exactly the same

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: I asked claude to give an example of how this could cause a difference and got this:

```
             tag: 2.0.0
                 ↓
    A ── B ─── M ── D  (HEAD)
         │    ╱
         └── C
             ↑
          tag: 1.5.0

  A is tagged 1.0.0

  You're on main. You branched off B to do some work (commit C, tagged 1.5.0), then merged it back at M, which you tagged 2.0.0. HEAD is one commit past that.

  Current code (first-parent walk: D → M → B → A): hits tags 2.0.0 (M), 1.0.0 (A). Result: [2.0.0, 1.0.0] — misses 1.5.0 because the walk never follows the merge into C.

  Proposed (for-each-ref --merged HEAD --sort=-version:refname): all tags reachable from HEAD = {1.0.0, 1.5.0, 2.0.0}. Sorted: [2.0.0, 1.5.0, 1.0.0].
```

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: But I think as we only tag develop we should not run into any issues

Comment (felixdesouza) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: there's also the whole thing with RCs and pre-merge

I think it's worth adding extra tests as a separate PR that codifies the *existing* behaviour, so we can be sure that we're not regressing here.

I would probably do that before this PR, or change this PR to not do any of the `GitOperations` stuff, but has the new tests and you just add some more to validate the behaviour you expect.

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: Okay - I'll put this on pause, try adding as many edge case tests as we can, then see if this works

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: I attempted the above suggestions in [743361a](https://github.com/palantir/gradle-revapi/pull/714/commits/743361aa645e668058a634e363cd5ccaed950737) and I think it much cleaner so hopefully it does!

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: I've added a bunch of tests here - https://github.com/palantir/gradle-revapi/pull/716 - 4 failed with the above approach, 1 I think is a genuine failure, 2 were due to the tests locking in topological order (even if tag dates where out of order etc) which I think maybe a bit unrealistic, and I think the last may arguably be a bug with current behaviour

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: okay so if we want to keep exact current behaviour but without the back walking I think something like this will work:

```java
    public final Provider<List<String>> previousGitTags() {
        return nonRootTags().zip(rootTag(), GitOperations::concat).zip(zeroZeroZeroHasParent(), GitOperations::filterAndLimit);
    }

    private Provider<List<String>> nonRootTags() {
        return getGitInvoker()
                .invokeWithResult(
                        "log",
                        "--topo-order",
                        "--simplify-by-decoration",
                        "--decorate-refs=refs/tags",
                        "--pretty=format:%D",
                        "--max-count=" + TAGS_TO_RETURN,
                        "HEAD^")
                .map(GitOperations::parseDecoratedTags);
    }

    // `--simplify-by-decoration` never emits the root commit
    private Provider<List<String>> rootTag() {
        return getGitInvoker()
                .invokeWithResult("log", "--max-parents=0", "--decorate-refs=refs/tags", "--pretty=format:%D", "HEAD^")
                .map(GitOperations::parseDecoratedTags);
    }
```

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: marked as a break incase this causes subtle changes, I think we will see no changes as we always squash merge

Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:63]: I've decided to go back to using `git describe` - I cannot find an alternative that behaves exactly the same and at the point that we are changing the behaviour we may as well shift entirely away from `git`. Which if we want todo can be a FLUP

File: src/main/java/com/palantir/gradle/revapi/GitOperations.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    private Provider<Boolean> commitExistsAt(String ref) {
+        return getGitInvoker()
+                .invokeWithResult("cat-file", "-t", ref)
+                .map(result -> "commit".equals(result.uncheckedStandardOut()));
+    }
+
+    private Provider<String> describeTagAt(String ref) {
+        return getGitInvoker()
+                .invokeWithResult("describe", "--tags", "--abbrev=0", ref)
```
Comment (crogoz) [inline src/main/java/com/palantir/gradle/revapi/GitOperations.java:83]: I know this is exactly what it used to be before, but checking the git common operations: https://github.com/palantir/gradle-git-version/blob/develop/src/main/java/com/palantir/gradle/gitversion/CommonGitOperations.java#L122 I see that we are using `--first-parent` when using the describe tags. I guess this could influence what tags we find for merge commits ? might be useful to add a test in the future for this to double check the behaviour ?

**Review by crogoz — APPROVED** [review]

File: src/main/java/com/palantir/gradle/revapi/GitOperations.java (hunk 8 above)
Comment (FinlayRJW) [reply src/main/java/com/palantir/gradle/revapi/GitOperations.java:83]: Yeah, `--first-parent` would slightly change the behaviour on non-squash merges. Honestly I'm not convinced the current behaviour is even "correct", describe finds the nearest reachable tag through the commit graph, which makes the answer depend on merge topology rather than on what was actually released.

I think resolving the last published version via Gradle is the right way forward, since module versions are the source of truth for what shipped. Range constraints do behave a bit weirdly around rc releases ([see here](https://github.com/palantir/gradle-revapi/compare/develop...finlayw/gradle-version-resolution#diff-4d690f153c2da54050494210e48f73068ac0a34a6eb90e4298908c1dbfc1ba7cR85-R94)), though I may have misused the range constraint API there.
